File: solutions/2026-04-07/solution.java
// Minimum Window Substring
// Difficulty: Hard
// URL: https://leetcode.com/problems/minimum-window-substring/

public class Solution {
    public String minWindow(String s, String t) {
        if (s == null || t == null || s.length() == 0 || t.length() == 0) {
            return "";
        }

        int[] tCount = new int[128];
        int[] sCount = new int[128];
        for (char c : t.toCharArray()) {
            tCount[c]++;
        }

        int required = t.length();
        int formed = 0;
        int[] windowCounts = new int[128];
        int l = 0;
        int minLen = Integer.MAX_VALUE;
        String minWindow = "";

        for (int r = 0; r < s.length(); r++) {
            char character = s.charAt(r);
            windowCounts[character]++;

            if (tCount[character] > 0 && windowCounts[character] == tCount[character]) {
                formed++;
            }

            while (l <= r && formed == required) {
                character = s.charAt(l);
                if (r - l + 1 < minLen) {
                    minLen = r - l + 1;
                    minWindow = s.substring(l, r + 1);
                }
                windowCounts[character]--;
                if (tCount[character] > 0 && windowCounts[character] < tCount[character]) {
                    formed--;
                }
                l++;
            }

        }
        return minWindow;
    }
}

File: solutions/2026-04-07/solution.py
# Minimum Window Substring
# Difficulty: Hard
# URL: https://leetcode.com/problems/minimum-window-substring/

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ''

        t_count = {}
        s_count = {}
        for char in t:
            t_count[char] = t_count.get(char, 0) + 1

        required = len(t_count)
        formed = 0
        window_counts = {}
        l = 0
        min_len = float('inf')
        min_window = ''

        for r in range(len(s)):
            character = s[r]
            window_counts[character] = window_counts.get(character, 0) + 1

            if character in t_count and window_counts[character] == t_count[character]:
                formed += 1

            while l <= r and formed == required:
                character = s[l]
                if r - l + 1 < min_len:
                    min_len = r - l + 1
                    min_window = s[l:r + 1]
                window_counts[character] -= 1
                if character in t_count and window_counts[character] < t_count[character]:
                    formed -= 1
                l += 1

        return min_window